File: replacesvgfillcolors/replace_svg_fillcolors.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');

// Default color mapping
const DEFAULT_COLOR_MAP = {
    '#ff0000': '#00ff00',    // Red to Green
    'red': 'blue',           // Named color
    'rgb(255, 0, 0)': 'rgb(0, 0, 255)',  // RGB format
    'rgba(255, 0, 0, 1)': 'rgba(0, 255, 0, 1)',  // RGBA format
    '#fff': '#000'           // Short hex
};

function replaceColors(svgContent, colorMap) {
    // Create a combined pattern to match all colors in the map
    const colorPatterns = Object.keys(colorMap).map(key => {
        // Escape special regex characters in the color string
        return key.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    }).join('|');

    // Regex that specifically looks for fill="color" or fill='color' or fill=color
    const regex = new RegExp(`(fill\\s*=\\s*["']?)(${colorPatterns})(["']?)`, 'gi');

    return svgContent.replace(regex, (match, prefix, color, suffix) => {
        // Find the replacement (case-sensitive first, then case-insensitive)
        let replacement = colorMap[color];
        if (!replacement) {
            const lowerColor = color.toLowerCase();
            for (const key in colorMap) {
                if (key.toLowerCase() === lowerColor) {
                    replacement = colorMap[key];
                    break;
                }
            }
        }

        // Return original if no replacement found
        return replacement ? `${prefix}${replacement}${suffix}` : match;
    });
}

function processSvgFile(inputPath, colorMap, outputSuffix = '_recolored') {
    try {
        const svgContent = fs.readFileSync(inputPath, 'utf8');
        const modifiedSvg = replaceColors(svgContent, colorMap);

        const parsedPath = path.parse(inputPath);
        const outputPath = path.join(
            parsedPath.dir,
            `${parsedPath.name}${outputSuffix}${parsedPath.ext}`
        );

        fs.writeFileSync(outputPath, modifiedSvg);
        console.log(`Modified SVG saved to: ${outputPath}`);
    } catch (error) {
        console.error('Error processing SVG file:', error.message);
        process.exit(1);
    }
}

// Main execution
if (require.main === module) {
    const fs = require('fs');
    const args = process.argv.slice(2);

    if (args.length < 1 || args[0] === '--help') {
        console.log('Usage: node svgColorReplacer.js <input-file> [<color-map-json>|--file <file.json>] [<output-suffix>]');
        console.log('  <input-file>     : The input SVG file with fill colors that should be replaced.')
        console.log('  <color-map-json> : The JSON color replacement dict as a string, properly shielded from shell expansion of special characters by quoting (single ticks for most shells). Example: \'{"#ff0000":"#00ff00","blue":"red"}\'')
        console.log('  <output-suffix>  : Optional. A suffix to append to the input file basename to construct the output file name. E.g., with input file \'robot.svg\' and suffix \'_red\', you will get output file \'robot_red.svg\'. Defaults to \'_recolored\'.');
        console.log('Examples:');
        console.log('  node svgColorReplacer.js input.svg \'{"#00ff00":"#ff0000","blue":"red"}\' _red');
        console.log('  node svgColorReplacer.js input.svg --file colors.json');
        console.log('  node svgColorReplacer.js input.svg --file colors.json _red');
        console.log('If you supply the color mapping via --file, the contents of that file should be just a single JSON dict like: \'{"#00ff00":"#ff0000","blue":"red"}\'')
        console.log()
        process.exit(0);
    }

    const inputPath = args[0];
    let colorMap = DEFAULT_COLOR_MAP;
    let outputSuffix = '_recolored';

    // Check for --file option
    const fileArgIndex = args.indexOf('--file');
    if (fileArgIndex !== -1) {
        // Read color map from file
        if (fileArgIndex + 1 >= args.length) {
            console.error('Error: --file option requires a filename');
            process.exit(1);
        }
        
        const colorMapFile = args[fileArgIndex + 1];
        try {
            const fileContent = fs.readFileSync(colorMapFile, 'utf8');
            colorMap = JSON.parse(fileContent);
        } catch (e) {
            console.error(`Error reading or parsing color map file: ${e.message}`);
            process.exit(1);
        }
    } else if (args[1] && !args[1].startsWith('--')) {
        // Parse color map from command line argument
        try {
            colorMap = JSON.parse(args[1]);
        } catch (e) {
            console.error('Invalid JSON color map provided. Using default.');
        }
    }

    // Get output suffix (skip the file args if they were used)
    const suffixArgIndex = fileArgIndex !== -1 ? fileArgIndex + 2 : 2;
    if (args[suffixArgIndex]) {
        outputSuffix = args[suffixArgIndex];
    }

    console.log(`Processing input file '${inputPath}' and adding output suffix '${outputSuffix}' to output file.`)
    console.log(`Using color map: ${JSON.stringify(colorMap, null, 2)}`)
    processSvgFile(inputPath, colorMap, outputSuffix);
}

module.exports = { replaceColors, processSvgFile };

File: replacesvgfillcolors/replace_svg_fillcolors.py
import sys
import re

def replace_colors(svg_content, color_map):
    """
    Replace fill colors in the given SVG content based on the color_map.

    Args:
        svg_content (str): The original SVG content.
        color_map (dict): A dictionary mapping old colors to new colors.

    Returns:
        str: The modified SVG content with replaced colors.
    """
    def replace_match(match):
        color = match.group(1)
        return f'fill="{color_map.get(color, color)}"'

    # Match fill="rgb(x, y, z)" and fill="#xxxxxx"
    pattern = r'fill="(rgb\(\d+,\s*\d+,\s*\d+\)|#[0-9a-fA-F]{6})"'
    return re.sub(pattern, replace_match, svg_content)

def main():
    if len(sys.argv) != 3:
        print("Usage: python replace_svg_colors.py <input_svg> <output_svg>")
        sys.exit(1)

    input_file = sys.argv[1]
    output_file = sys.argv[2]

    # Define the color replacement mapping
    color_map = {
        "rgb(124, 190, 178)": "rgb(172, 225, 217)",  # Pastel teal
        "rgb(217, 167, 81)": "rgb(242, 210, 151)",  # Pastel orange
        "gray": "rgb(211, 211, 211)",  # Light gray (softened)
        "white": "rgb(255, 255, 255)",  # White (unchanged)
        "rgb(200, 200, 200)": "rgb(230, 230, 230)",  # Lighter gray
        "rgb(148, 167, 203)": "rgb(190, 200, 230)"  # Pastel blue
    }


    try:
        with open(input_file, "r", encoding="utf-8") as f:
            svg_content = f.read()

        modified_svg = replace_colors(svg_content, color_map)

        with open(output_file, "w", encoding="utf-8") as f:
            f.write(modified_svg)

        print(f"Updated SVG saved to {output_file}")

    except FileNotFoundError:
        print(f"Error: File '{input_file}' not found.")
        sys.exit(1)
    except Exception as e:
        print(f"An error occurred: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()
